modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/ParsingPhaseHandler.java

```diff
@@ -0,0 +1,95 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.sql.engine.exec.fsm;
+
+import static org.apache.ignite.internal.sql.engine.QueryProperty.ALLOWED_QUERY_TYPES;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import org.apache.ignite.internal.sql.engine.SqlQueryType;
+import org.apache.ignite.internal.sql.engine.property.SqlProperties;
+import org.apache.ignite.internal.sql.engine.sql.ParsedResult;
+
+/** Parses the query string and populate {@link Query query state} with results. */
+class ParsingPhaseHandler implements ExecutionPhaseHandler {
+    static final ExecutionPhaseHandler INSTANCE = new ParsingPhaseHandler();
+
+    private ParsingPhaseHandler() { }
+
+    @Override
+    public Result handle(Query query) {
+        ParsedResult parsedResult = query.executor.lookupParsedResultInCache(query.sql);
+
+        if (parsedResult != null) {
+            query.parsedResult = parsedResult;
+
+            query.moveTo(ExecutionPhase.OPTIMIZING);
+
+            return Result.proceedImmediately();
+        }
+
+        CompletableFuture<Void> awaitFuture = new CompletableFuture<>();
+        query.executor.execute(() -> {
+            try {
+                ParsedResult result;
+                if (multiStatementQueryAllowed(query.properties)) {
+                    List<ParsedResult> results = query.executor.parseScript(query.sql);
+
+                    if (results.size() != 1 || results.get(0).queryType() == SqlQueryType.TX_CONTROL) {
+                        // the query is script indeed, need different execution path
+                        query.parsedScript = results;
+
+                        query.moveTo(ExecutionPhase.SCRIPT_INITIALIZATION);
+                        awaitFuture.complete(null);
+
+                        return;
+                    }
+
+                    result = results.get(0);
+                } else {
+                    result = query.executor.parse(query.sql);
+                }
+
+                if (shouldBeCached(result.queryType())) {
+                    query.executor.updateParsedResultCache(query.sql, result);
+                }
+
+                query.parsedResult = result;
+
+                query.moveTo(ExecutionPhase.OPTIMIZING);
+                awaitFuture.complete(null);
+            } catch (Throwable th) {
+                awaitFuture.completeExceptionally(th);
+            }
+        });
+
+        return Result.proceedAfter(awaitFuture);
+    }
+
+    private static boolean shouldBeCached(SqlQueryType queryType) {
+        return queryType == SqlQueryType.QUERY || queryType == SqlQueryType.DML;
+    }
+
+    /** Returns {@code true} if the specified properties allow multi-statement query execution. */
+    private static boolean multiStatementQueryAllowed(SqlProperties properties) {
```

Comment:
Perhaps we can use `Commons::isMultiStatementQueryAllowed` here ?

Reply:
good catch! Replaced